src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerations/AvoidMultipleEnumerations.Helpers.cs

```diff
@@ -0,0 +1,274 @@
+                            && parameters[0].Type.OriginalDefinition.SpecialType == SpecialType.System_Collections_Generic_IEnumerable_T)
+                        {
+                            builder.AddRange(methodSymbol);
+                        }
+                    }
+                }
+            }
+
+            return builder.ToImmutable();
+        }
+
+        private static bool IsDeferredType(ITypeSymbol type, ImmutableArray<ITypeSymbol> additionalTypesToCheck)
+            => type.SpecialType is SpecialType.System_Collections_Generic_IEnumerable_T or SpecialType.System_Collections_IEnumerable
+               || additionalTypesToCheck.Contains(type);
+
+        private static ImmutableArray<ITypeSymbol> GetTypes(Compilation compilation, ImmutableArray<string> typeNames)
+        {
+            using var builder = ArrayBuilder<ITypeSymbol>.GetInstance();
+            foreach (var name in typeNames)
+            {
+                if (compilation.TryGetOrCreateTypeByMetadataName(name, out var typeSymbol))
+                {
+                    builder.Add(typeSymbol);
+                }
+            }
+
+            return builder.ToImmutable();
+        }
+
+        private static ImmutableArray<IMethodSymbol> GetTwoParametersEnumeratedMethods(WellKnownTypeProvider wellKnownTypeProvider)
+        {
+            using var builder = ArrayBuilder<IMethodSymbol>.GetInstance();
+            GetWellKnownMethods(wellKnownTypeProvider, WellKnownTypeNames.SystemLinqEnumerable, s_linqTwoParametersEnumeratedMethods, builder);
+            return builder.ToImmutable();
+        }
+
+        private static ImmutableArray<IMethodSymbol> GetOneParameterDeferredMethods(WellKnownTypeProvider wellKnownTypeProvider)
+        {
+            using var builder = ArrayBuilder<IMethodSymbol>.GetInstance();
+            GetWellKnownMethods(wellKnownTypeProvider, WellKnownTypeNames.SystemLinqEnumerable, s_linqOneParameterDeferredMethods, builder);
+            return builder.ToImmutable();
+        }
+
+        private static ImmutableArray<IMethodSymbol> GetTwoParametersDeferredMethods(WellKnownTypeProvider wellKnownTypeProvider)
+        {
+            using var builder = ArrayBuilder<IMethodSymbol>.GetInstance();
+            GetWellKnownMethods(wellKnownTypeProvider, WellKnownTypeNames.SystemLinqEnumerable, s_linqTwoParametersDeferredMethods, builder);
+            return builder.ToImmutable();
+        }
+
+        private static void GetWellKnownMethods(
+            WellKnownTypeProvider wellKnownTypeProvider,
+            string typeName,
+            ImmutableArray<string> methodNames,
+            ArrayBuilder<IMethodSymbol> builder)
+        {
+            if (wellKnownTypeProvider.TryGetOrCreateTypeByMetadataName(typeName, out var type))
+            {
+                foreach (var methodName in methodNames)
+                {
+                    builder.AddRange(type.GetMembers(methodName).Cast<IMethodSymbol>());
```

Comment:
It would be more safer to replace `.Cast<IMethodSymbol>()` with `.OfType<IMethodSymbol>()`

Reply:
Done